neuralpp/symbolic/basic_expression.py

```diff
@@ -1,25 +1,68 @@
 from __future__ import annotations
-from neuralpp.symbolic.expression import Expression, FunctionApplication, Variable, Constant, AtomicExpression
+
+import operator
+import builtins
+from neuralpp.symbolic.expression import Expression, FunctionApplication, Variable, Constant, AtomicExpression, \
+    VariableNotTypedError, FunctionType
 from abc import ABC
-from typing import Any, List, Type
+from typing import Any, List, Optional, Type, Callable, Tuple
+
+
+class AmbiguousTypeError(TypeError, ValueError):
+    def __init__(self, python_callable: Callable):
+        super().__init__(f"{python_callable} is ambiguous.")
+
+
+def infer_python_callable_type(python_callable: Callable) -> FunctionType:
+    match python_callable:
+        # boolean operation
+        case operator.__and__ | operator.__or__ | operator.__xor__:
+            raise AmbiguousTypeError(python_callable)  # this is also ambiguous because the arity could be arbitrary
+            # return FunctionType([bool, bool], bool)
+        case operator.__not__:
+            return FunctionType([bool], bool)
+        # comparison
+        case operator.le | operator.lt | operator.ge | operator.gt | operator.eq:
+            raise AmbiguousTypeError(python_callable)
+        # arithmetic
+        case operator.add | operator.mul | operator.pow:
+            raise AmbiguousTypeError(python_callable)
+        # min/max
+        case builtins.min | builtins.max:
+            raise AmbiguousTypeError(python_callable)
+        case _:
+            raise ValueError(f"Python callable {python_callable} is not recognized.")
 
 
 class BasicExpression(Expression, ABC):
     @classmethod
-    def new_constant(cls, value: Any) -> BasicConstant:
-        return BasicConstant(value)
+    def new_constant(cls, value: Any, constant_type: Optional[Expression] = None) -> BasicConstant:
+        return BasicConstant(value, constant_type)
 
     @classmethod
-    def new_variable(cls, name: str) -> BasicVariable:
-        return BasicVariable(name)
+    def new_variable(cls, name: str, variable_type: Expression) -> BasicVariable:
+        return BasicVariable(name, variable_type)
 
     @classmethod
     def new_function_application(cls, function: Expression, arguments: List[Expression]) -> BasicFunctionApplication:
         return BasicFunctionApplication(function, arguments)
 
 
+def new_type(python_type: Type) -> Expression:
+    """ A handy method to create properly wrapped type as an Expression. E.g., new_type(int). """
```

Comment:
nit: the terms "convenience method" or "helper methods" are more standard.

Reply:
this function is gone now, but noted.